src/bot_backend.py

```diff
@@ -183,14 +189,25 @@ def add_file_message(self, path, bot_msg):
         )
 
     def add_function_call_response_message(self, function_response: str, save_tokens=True):
+        # Add code to notebook cells.
```

Comment:
When trying to import either `get_code_str` or `parse_json`, I run into `ImportError` due to circular import.
This is caused by  `from bot_backend.py import *` in `functional.py` and `from functional import *` in response_parser.py` .
Since `response_parser` imports `functional` and `functional` imports `bot_backend` importing either function/method results in a circular import.

I would suggest moving `parse_json` into `response_parser` since it is the only place where it is used and replace from import.

As a quick fix I am importing `parse_json` in the method `add_function_call_response_message` itself.

Reply:
You can use `from functional import parse_json` instead of importing the entire module.

Reply:
Sorry about that, this cannot solve the `ImportError` issue. Please use your quick fix first. I will reorganize my code as soon as I have time.

Reply:
I used `parse_json` but when `self.function_args_str` is not in json format, calling `parse_json` raises an exception:
```
Traceback (most recent call last):
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/routes.py", line 442, in run_predict
    output = await app.get_blocks().process_api(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/blocks.py", line 1392, in process_api
    result = await self.call_function(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/blocks.py", line 1111, in call_function
    prediction = await utils.async_iteration(iterator)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 346, in async_iteration
    return await iterator.__anext__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 339, in __anext__
    return await anyio.to_thread.run_sync(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 322, in run_sync_iterator_async
    return next(iterator)
           ^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 691, in gen_wrapper
    yield from f(*args, **kwargs)
  File "/home/mauro/projects/local_code_interpreter_fork/src/web_ui.py", line 112, in bot
    exit(-1)
  File "<frozen _sitebuiltins>", line 26, in __call__
SystemExit: -1
```
Which is weird because I have a try/except to catch any error:
```python
        from functional import parse_json
        try:
            code_str = parse_json(self.function_args_str, True)
            if code_str:
                add_code_cell_to_notebook(code_str)
        except Exception as e:
            print(f"Caught exception while trying to add code to notebook: {e}")
```

On the other hand I never get any issue when using `json.loads`.
I see in the docstring of `parse_json` that the reason for the implementation of this warper around `json.loads` are the 
`\\n`s in the string but this issue as been solved with the latest `1106` versions of GPT chat completion.

I faced a similar issue when using the older versions in previous projects.
To prevent this potential case where the string is in *erronous* json, I suggest we simply apply a `replace('\\n', '\n')`
on `function_args_str` before callin `json.loads` and leave the rest as is.

What do you say?

Reply:
In the following code, `s` is an erroneous JSON sting that I encountered in a previous test. However, applying `replace('\\n', '\n')` to it alone doesn't allow successful parsing using `json.loads`.

```python
import json

s = '''{
    "code": "
import numpy as np

x = np.random.rand(100, 100)
y = np.random.rand(100, 100)

z = np.dot(x, y)
z
"
}'''

result = json.loads(s.replace('\\n', '\n'))
```

To minimize code duplication, I've added a `code_str` attribute to the `BotBackend` class. You can rebase your branch and use `self.code_str` to obtain the code string.